ios/RNZendeskChatModule.m

```diff
@@ -201,4 +213,84 @@ - (void) dismissChatUI {
 	});
 }
 
+RCT_EXPORT_METHOD(sendMessage: (NSString*)message
+		resolver:(RCTPromiseResolveBlock)resolve
+		rejecter:(RCTPromiseRejectBlock)reject
+	) {
+		dispatch_sync(dispatch_get_main_queue(), ^{
+			[ZDKChat.chatProvider sendMessage:message completion:^(NSString *messageId, NSError *error) {
+				if (error) {
+					reject(@"error", @"no message id returned", error);
+				} else {
+					resolve(messageId);
+				}
+			}];
+		});
+}
+
+RCT_EXPORT_METHOD(sendChatRating: (NSString*)rating
+		resolver:(RCTPromiseResolveBlock)resolve
+		rejecter:(RCTPromiseRejectBlock)reject
+	) {
+        ZDKRating convertedRating = [RCTConvert ZDKRating:rating];
+
+		dispatch_sync(dispatch_get_main_queue(), ^{
+			[ZDKChat.chatProvider sendChatRating:convertedRating completion:^(ZDKRating _rating, NSError *error) {
+				if (error) {
+					reject(@"error", @"no rating returned", error);
+				} else {
+					resolve(rating);
+				}
+			}];
+		});
+}
+
+RCT_EXPORT_METHOD(endChat: (RCTPromiseResolveBlock)resolve
+		rejecter:(RCTPromiseRejectBlock)reject
+	) {
+		dispatch_sync(dispatch_get_main_queue(), ^{
+			[ZDKChat.chatProvider endChat:^(BOOL didEndChat, NSError *error) {
+				if (didEndChat) {
+					resolve(didEndChat ? @"true" : @"false");
+				} else {
+					reject(@"error", @"chat no ended", error);
+				}
+			}];
+		});
+}
+
+RCT_EXPORT_METHOD(_sendOfflineFormWith3Args:(NSString *)message visitorInfo:(NSDictionary *)visitorInfo departmentId:(NSString *)departmentId
+		resolver:(RCTPromiseResolveBlock)resolve
+		rejecter:(RCTPromiseRejectBlock)reject
+	) {
+
+		dispatch_sync(dispatch_get_main_queue(), ^{
+			ZDKVisitorInfo *visitor = nil;
+
+			if (!message) {
+				reject(@"error", @"message not provided", nil);
+				return;
+			}
+
+			if (visitorInfo) {
+				visitor = [[ZDKVisitorInfo alloc] initWithName:visitorInfo[@"name"]
+                                                  email:visitorInfo[@"email"]
+												  phoneNumber:visitorInfo[@"phone"]];
+			}
+
+			ZDKOfflineForm *form = [[ZDKOfflineForm alloc]  initWithVisitorInfo:visitor
+                                                  	  departmentId:departmentId
+                                                      message:message];
+
+			[ZDKChat.chatProvider sendOfflineForm:form
+                           completion:^(ZDKOfflineForm *offlineForm, NSError *error) {
+				  if (error) {
+						reject(@"error", @"offline form couln't be sent", error);
+					} else {
+						resolve(offlineForm);
```

Comment:
I don't know why I'm able to return `offlineForm`... but it's wasn't showing up any error, so there you have it :)

Reply:
Resolve isn't type-safe -- are you sure this is what you want to do?

Reply:
definitely was a wrong idea, I was just expecting to see a compiler error or something... I don't understand how static typing languages allows these things... 

So I should just return `@"true"` or something like that?

Reply:
You don't have to to pass anything into `resolve()` unless you have a purpose for it.

Reply:
And re: static languages, Objective-C has a highly dynamic runtime-typed language nature to it. It has many of the benefits of Static languages -- unless you opt-out.

 `resolve()` definitely opts-out, since it's a library function meant to be used everywhere, and there's no "automatic" way for it to know which objects are valid or invalid for the JS.